inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/gather_nd_ref.cl

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2021 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "include/fetch.cl"
+
+#define GET_UPDATES_INDEX(prefix, idx_order) CAT(prefix, _GET_INDEX)(idx_order)
+#define GET_OUTPUT_INDEX(out_order) OUTPUT_GET_INDEX(out_order)
+
+#if INPUT0_DIMS == 4
+    #define IN_ORDER in_b,in_f,in_y,in_x
+#elif INPUT0_DIMS == 5
+    #define IN_ORDER in_b,in_f,in_z,in_y,in_x
+#elif INPUT0_DIMS == 6
+    #define IN_ORDER in_b,in_f,in_w,in_z,in_y,in_x
+#endif
+
+#if INPUT1_DIMS == 4
+    #define IDX_ORDER idx_b,idx_f,idx_y,idx_x
+#elif INPUT1_DIMS == 5
+    #define IDX_ORDER idx_b,idx_f,idx_z,idx_y,idx_x
+#elif INPUT1_DIMS == 6
+    #define IDX_ORDER idx_b,idx_f,idx_w,idx_z,idx_y,idx_x
+#endif
+
+#if OUTPUT_DIMS == 4
+    #define OUT_ORDER out_b,out_f,out_y,out_x
+#elif OUTPUT_DIMS == 5
+    #define OUT_ORDER out_b,out_f,out_z,out_y,out_x
+#elif OUTPUT_DIMS == 6
+    #define OUT_ORDER out_b,out_f,out_w,out_z,out_y,out_x
+#endif
+
+#define INDICES_MAX_DIM 6
+
+KERNEL(gather_nd_ref)(const __global INPUT0_TYPE* data,
+                   const __global INPUT1_TYPE* indices,
+                   __global OUTPUT_TYPE* output
+#if HAS_FUSED_OPS_DECLS
+                   , FUSED_OPS_DECLS
+#endif
+)
+{
+
+    const uint dim0 = get_global_id(0);
+    const uint dim1 = get_global_id(1);
+    const uint dim2 = get_global_id(2);
+
+    //Calculate indice index
+    const uint F_NUM = (INDICES_RANK == 2) ? 1 : INPUT1_FEATURE_NUM;
+    const uint idx_f = dim2 % F_NUM;
+    const uint idx_b = dim2 / F_NUM;
+
+    #if INPUT1_DIMS == 4
+        const uint idx_x = dim0;
+        const uint idx_y = dim1;
+        const uint idx_z = 0;
+        const uint idx_w = 0;
+
+        const uint idx_arr[INPUT1_DIMS*2] = {idx_b, idx_f, idx_y, idx_x, 0, 0, 0, 0};
+        const uint idx_dim[INPUT1_DIMS] = {INPUT1_BATCH_NUM, INPUT1_FEATURE_NUM, INPUT1_SIZE_Y, INPUT1_SIZE_X};
+    #elif INPUT1_DIMS == 5
+        const uint X_NUM = (INDICES_RANK == 5) ? 1 : INPUT1_SIZE_X;
+
+        const uint idx_x = dim0 % X_NUM;
+        const uint idx_y = dim0 / X_NUM;
+        const uint idx_z = dim1;
+        const uint idx_w = 0;
+
+        const uint idx_arr[INPUT1_DIMS*2] = {idx_b, idx_f, idx_z, idx_y, idx_x, 0, 0, 0, 0, 0};
+        const uint idx_dim[INPUT1_DIMS] = {INPUT1_BATCH_NUM, INPUT1_FEATURE_NUM, INPUT1_SIZE_Z, INPUT1_SIZE_Y, INPUT1_SIZE_X};
+    #elif INPUT1_DIMS == 6
```

Comment:
Why do you use "elif" construction here? If there may be another INPUT1_DIMS cases then unexpected result will be here. My point is that we should use "else" construction in cases when all possible variants are under "if" or use "#error" message after `#elif INPUT1_DIMS == 6` when there may be something else.

Reply:
Done.

Reply:
Oh, as I can see, we have all input and output dimenshion checks in code lines from 20 to 42. What do you think, may be, it'll be better to add 3 preprocessing "#error" directives here (into lines 20 - 42). And after that in the kernel body to use such construction as: 
```
#if DIMS_NUM == 4
...
#elif DIMS_NUM == 5
...
#else
...
#endif
```

In this case all unsupported dimension checks will be in a one place before the kernel body. I'm sorry that I didn't see it in my first review attempt...

Reply:
How about DIMS_NUM == 2 or 3? I think DIMS_NUM < 4 case should go else(#error). So the #elif DIMS_NUM ==6 and #else are need.

Reply:
You can process all these variants here:

```
#if INPUT0_DIMS == 4
    #define IN_ORDER in_b,in_f,in_y,in_x
#elif INPUT0_DIMS == 5
    #define IN_ORDER in_b,in_f,in_z,in_y,in_x
#elif INPUT0_DIMS == 6
    #define IN_ORDER in_b,in_f,in_w,in_z,in_y,in_x
#else
#error gather_nd_ref.cl - not supported INPUT0_DIMS_NUM
#endif
```

And the same for INPUT1_DIMS and OUTPUT_DIMS. As for me, this will simplify your code in the kernel body. You'll may use only:

```
#if INPUT1_DIMS == 4
   ...
#elif INPUT1_DIMS == 5
   ...
#else
   ...
#endif
```

And all checks for DIMS NUM will be placed before the kernel body.

Reply:
Okay, I made the code simple as you your suggestion.

Reply:
I thought that you'll make error processing before the kernel body but I suppose your variant also'll work corectly because in GetSupportedKeys method we have only simple bfyx, bfzyx, bfwzyx formats for input and output layouts. Thanks.